### PR TITLE
Enable basic auth on gateway

### PR DESCRIPTION
**Pull request by alexellis**
Signed-off-by: Alex Ellis (VMware) <alexellis2@gmail.com>

## Description

Fixes https://github.com/openfaas/faas/issues/687 allowing the
gateway to handle the responsibility of basic auth for when it is
in use.

To enable set basic_auth env-var to true and then mount two
secrets or plaintext files under /var/secrets/

basic_auth_user, basic_auth_password

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

#687

## How Has This Been Tested?

Tested with faas-cli list/deploy and with Safari browser.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: gateway/server.go [hunk 1]
```diff
@@ -33,6 +35,27 @@ func main() {
 
 	log.Printf("Binding to external function provider: %s", config.FunctionsProviderURL)
 
+	var credentials *handlers.BasicAuthCredentials
+
+	if config.UseBasicAuth {
+		userPath := "/var/secrets/basic_auth_user"
```
Comment (johnmccabe) [inline gateway/server.go]: You need to check the `/run/secrets/*` files when running on Swarm or the gateway will fail to initialise.

Comment (alexellis) [reply gateway/server.go]: In which case should I make the secret path a config item too?

Comment (johnmccabe) [reply gateway/server.go]: Could just check both, its only going to occur once on startup so shouldn't be an issue? Saves having to expose more config to the user.

Comment (alexellis) [reply gateway/server.go]: @stefanprodan is there such a thing as a standard location for secrets on Kubernetes?

Comment (stefanprodan) [reply gateway/server.go]: There is no  standard location, usually ok Kubernetes you would mount secrets as ENV vars not files on disk.

**Comment by alexellis** [comment]
@johnmccabe I think I've addressed your concern now. It is not on by default since we need to figure out how to have the user create the secrets first.

File: docker-compose.yml [hunk 2]
```diff
@@ -16,6 +16,8 @@ services:
             faas_nats_port: 4222
             direct_functions: "true"    # Functions are invoked directly over the overlay network
             direct_functions_suffix: ""
+            basic_auth: "false"
```
[the remark this answers is not in the document]
Comment (alexellis) [reply docker-compose.yml]: I think to use it you would need to do that, but we'd not be able to do a stack deploy unless these secrets existed.

**Comment by alexellis** [comment]
Thank you guys for reviewing and for your feedback.

**Pull request merged**